calyx-py/test/correctness/queues/binheap.py

```diff
@@ -29,107 +31,369 @@ def insert_swap(prog, name, width, len, idx_w):
     return comp
 
 
-def insert_binheap(prog, name):
+def insert_binheap(prog, name, factor):
     """Inserts the component `binheap` into the program.
 
     It is a minimum binary heap, represented as an array.
 
     The heap just supports the `push` operation.
-    Its only input is `value`, the value to push to the queue.
+    Its only inputs are `value` and `rank`, the value and rank to push to the queue.
     """
-    comp: cb.ComponentBuilder = prog.component(name)
-    value = comp.input("value", 64)  # The value to push to the heap
 
-    swap = comp.cell("swap", insert_swap(prog, "swap", 64, 15, 4))
+    comp = prog.component(name)
+    
+    n = (2**factor) - 1
+
+    cmd = comp.input("cmd", 2)
+    # If it is 0, we pop.
+    # If it is 1, we peek.
+    # If it is 2, we push `(rank, value)` to the queue.
 
-    mem = comp.seq_mem_d1("mem", 64, 15, 4, is_ref=True)
+    # The rank and value to push to the heap.
+    rank = comp.input("rank", 32)
+    value = comp.input("value", 32)
+
+    swap = comp.cell("swap", insert_swap(prog, "swap", 64, n, factor))
+    tuplify = comp.cell("tuplify", insert_tuplify(prog, "tuplify", 32, 32))
+    untuplify = comp.cell("untuplify", insert_untuplify(prog, "untuplify", 32, 32))
+
+    mem = comp.comb_mem_d1("mem", 64, n, factor)
     # The memory to store the heap, represented as an array.
-    # For now it has a hardcoded max length of 15, i.e., a binary heap of height 4.
-    # Each cell of the memory is 64 bits wide.
+    # Each cell of the memory is 64 bits wide: 32 for both rank and value.
```

Comment:
The comment may confuse some. It may be clearer to say something like:
> `# Each cell of the memory is 64 bits wide: a 32-bit rank and a 23-bit value.`